docs(CHANGELOG) changes for 2.6.0 (#7873)
Kong/kong: CHANGELOG.md

  [#7658](https://github.com/Kong/kong/pull/7658)
- Bumped `grpcurl` from 1.8.1 to 1.8.2
  [#7659](https://github.com/Kong/kong/pull/7659)
- Bumped `luasec` from 1.0.1 to 1.0.2
  [#7750](https://github.com/Kong/kong/pull/7750)
- Bumped `lua-resty-ipmatcher` to 0.6.1
  [#7703](https://github.com/Kong/kong/pull/7703)
  Thanks [EpicEric](https://github.com/EpicEric) for the patch!

All Kong Gateway OSS plugins will be moved from individual repositories and centralized
into the main Kong Gateway (OSS) repository. We are making a gradual transition. On this
release:

- Moved AWS-Lambda inside the Kong repo
  [#7464](https://github.com/Kong/kong/pull/7464).
- Moved ACME inside the Kong repo
  [#7464](https://github.com/Kong/kong/pull/7464).
- Moved Prometheus inside the Kong repo
  [#7666](https://github.com/Kong/kong/pull/7666).
- Moved Session inside the Kong repo
  [#7738](https://github.com/Kong/kong/pull/7738).
- Moved GRPC-web inside the Kong repo
  [#7782](https://github.com/Kong/kong/pull/7782).
- Moved Serverless functions inside the Kong repo
  [#7792](https://github.com/Kong/kong/pull/7792).

### Additions

#### Core

- New schema entity validator: `mutually_exclusive`.

>   [#7765](https://github.com/Kong/kong/pull/7765)

Reviewer: @bungle could we have more details in the changelog maybe describing schema validators and about the `mutually_exclusive` validation type?

#### Performance

On this release we've done some special efforts with regards to performance.

There's a new performance workflow which periodically checks new code additions against some
typical scenarios [#7030](https://github.com/Kong/kong/pull/7030) [#7547](https://github.com/Kong/kong/pull/7547)

In addition to that, the following changes were specifically included to improve performance:

- Reduced unnecessary reads of `ngx.var`
  [#7840](https://github.com/Kong/kong/pull/7840)
- Loaded more indexed variables
  [#7849](https://github.com/Kong/kong/pull/7849)
- Optimized table creation in Balancer
  [#7852](https://github.com/Kong/kong/pull/7852)
- Reduce calls to `ngx.update_time`
  [#7853](https://github.com/Kong/kong/pull/7853)
- Use read-only replica for PostgreSQL meta-schema reading
  [#7454](https://github.com/Kong/kong/pull/7454)

> - URL escaping detects cases when it's not needed and early-exists

Reviewer 2: ```suggestion
- URL escaping detects cases when it's not needed and early-exits
```

  [#7742](https://github.com/Kong/kong/pull/7742)
- Accelerated variable loading via indexes
  [#7818](https://github.com/Kong/kong/pull/7818)

#### Configuration

> - Enable IPV6 on `dns_order` as unsupported experimental feature

Reviewer 2: ```suggestion
- Enable IPV6 on `dns_order` as unsupported experimental feature. Please
  give it a try and report back any issues.
```

>   [#7819](https://github.com/Kong/kong/pull/7819).

Reviewer 3: @Tieske mentioned this is a experiemnting feature, maybe we should also hide it from
changelog? (https://github.com/Kong/docs.konghq.com/pull/3253)

Reviewer 2: Since its clearly marked as experimental, I think it's fine here. Also because we want people to know it, so they can try and report issues

- The template renderer can now use `os.getenv`
  [#6872](https://github.com/Kong/kong/pull/6872).

#### Hybrid Mode

- Data plane is able to eliminate some unknown fields when Control Plane is using a more modern version
  [#7827](https://github.com/Kong/kong/pull/7827).

#### Admin API

> - Add support for HEAD requests

Reviewer: > Added the support for the HTTP HEAD method for all Admin API endpoints

> - Add support for HEAD requests

Reviewer: ```suggestion
- Added the support for the HTTP HEAD method for all Admin API endpoints
```

  [#7796](https://github.com/Kong/kong/pull/7796)
- Improve support for OPTIONS requests

>   [#7830](https://github.com/Kong/kong/pull/7830)

Reviewer: > Added better support for OPTIONS requests. Previously, the Admin API replied the same on all OPTIONS requests, where as now OPTIONS request only replies to routes that our Admin API has. Non-existing routes will have a 404 returned. It also adds Allow header to responses, and both Allow and Access-Control-Allow-Methods now contain only the methods that the specific API supports. #7830

> - Improve support for OPTIONS requests
>   [#7830](https://github.com/Kong/kong/pull/7830)

Reviewer: ```suggestion
- Added better support for OPTIONS requests. Previously, the Admin API replied the same on all OPTIONS requests, where as now OPTIONS request will only reply to routes that our Admin API has. Non-existing routes will have a 404 returned. It also adds Allow header to responses, and both Allow and Access-Control-Allow-Methods now contain only the methods that the specific API supports. [#7830](https://github.com/Kong/kong/pull/7830)
```

#### Plugins

> - **AWS-Lambda**: The plugin will now try to detect the AWS region by using `AWS_REGION` and

Reviewer: >  **AWS-Lambda:** The plugin will now try to detect the AWS region by using AWS_REGION and AWS_DEFAULT_REGION environment variables (when not specified with the plugin configuration) When invoking the Lambda, Kong passes the retrieved region, hence always invoking the Lamda in the same region where Kong is located. . #7765 

@bungle what do you think of this improved changelog entry?

>   `AWS_DEFAULT_REGION` environment variables (when not specified with the plugin configuration).

Reviewer 2: ```suggestion
  `AWS_DEFAULT_REGION` environment variables (when not specified with the plugin configuration).
  This allows to specify a 'region' on a per Kong node basis, hence adding the ability to invoke the
  Lamda in the same region where Kong is located.
```

  [#7765](https://github.com/Kong/kong/pull/7765)

> - **Datadog**: `host` and `port` config options can be configured from `ENV` variables

Reviewer: > Datadog: host and port config options can now be configured from ENV variables - which is necessary for Kubernetes installations that run the DataDog agent as a Daemon-Set.   #7463 Thanks rallyben for the patch!

@Tieske Something like this entry to enhance our 2.6 changelog? Should I add any color?

Reviewer 2: lemme add some

> - **Datadog**: `host` and `port` config options can be configured from `ENV` variables
>   [#7463](https://github.com/Kong/kong/pull/7463)

Reviewer 2: ```suggestion
- **Datadog**: `host` and `port` config options can be configured from environment variables
  `KONG_DATADOG_AGENT_HOST` and `KONG_DATADOG_AGENT_PORT`. This allows to set different
  destinations on a per Kong node basis, which makes multi-DC setups easier and in Kubernetes allows to
  run the datadog agents as a daemon-set.
  [#7463](https://github.com/Kong/kong/pull/7463)
```
wdyt @PaulFischer-Kong

  Thanks [rallyben](https://github.com/rallyben) for the patch!

> - **Prometheus**: The new `data_plane_cluster_cert_expiry_timestamp` exposes DP cert expiry timestamp

Reviewer: > Prometheus: A new metric `data_plane_cluster_cert_expiry_timestamp` is added to expose the Data Plane's `cluster_cert` expiry timestamp for improved monitoring in Hybrid Mode. #7800.

Added some more details to this entry.

>   [#7800](https://github.com/Kong/kong/pull/7800).

Reviewer: ```suggestion
- **Prometheus:** A new metric `data_plane_cluster_cert_expiry_timestamp` is added to expose the Data Plane's cluster_cert expiry timestamp for improved monitoring in Hybrid Mode. [#7800](https://github.com/Kong/kong/pull/7800).

**GRPC-Gateway**:

- The plugin can decode entities of type `.google.protobuf.Timestamp`

>   [#7538](https://github.com/Kong/kong/pull/7538)

Reviewer: @javierguerragiraldez could you improve these two changelog entries for the GRPC-Gateway section to include more details describing the addition?

> - The plugin can decode entities of type `.google.protobuf.Timestamp`
>   [#7538](https://github.com/Kong/kong/pull/7538)

Reviewer 4: would this be better?
```suggestion
- Fields of type `.google.protobuf.Timestamp` on the gRPC side are now
  transcoded to and from ISO8601 strings in the REST side.
  [#7538](https://github.com/Kong/kong/pull/7538)
```

> - Added support for structured URI arguments
>   [#7564](https://github.com/Kong/kong/pull/7564)

Reviewer 4: similarly
```suggestion
- URI arguments like `..?foo.bar=x&foo.baz=y` are interpreted as structured
  fields, equivalent to `{"foo": {"bar": "x", "baz": "y"}}`
  [#7564](https://github.com/Kong/kong/pull/7564)
```

  Thanks [git-torrent](https://github.com/git-torrent) for the patch!

**Request Termination**:

> - New `trigger` config option, which makes the plugin activate for any requests with a header or query parameter

Reviewer: @Tieske Would you have a moment to draft some short text for this changelog explaining the request termination plugin and excellent debugging aid use cases you mentioned in the PR? That would level up our 2.6 changelog.

Reviewer 2: added suggestions (2) in text below

Reviewer: @Tieske ty vm! Would you agree we should move this section 'request termination' under the plugins section, instead of its separate call out?

Reviewer 2: yes, that would make sense

> - New `trigger` config option, which makes the plugin activate for any requests with a header or query parameter
>   named like the trigger

Reviewer 2: ```suggestion
- New `trigger` config option, which makes the plugin only activate for any requests with a header or query parameter
  named like the trigger. This can be a great debugging aid, without impacting actual traffic being processed.
```

  [#6744](https://github.com/Kong/kong/pull/6744).

> - The `request-echo` config option. If not set, the plugin activates on all requests.
>   If set, then the plugin only activates for any requests with a header or query parameter
>   named like the trigger
>   [#6744](https://github.com/Kong/kong/pull/6744).

Reviewer 2: ```suggestion
- The `request-echo` config option. If set, the plugin responds with a copy of
  the incoming request. This eases troubleshooting when Kong is behind one or more
  other proxies or LB's, especially when combined with the new 'trigger' option.
  [#6744](https://github.com/Kong/kong/pull/6744).
```

#### Plugins

- **AWS-Lambda**: The plugin will now try to detect the AWS region by using `AWS_REGION` and
  `AWS_DEFAULT_REGION` environment variables (when not specified with the plugin configuration).
  This allows to specify a 'region' on a per Kong node basis, hence adding the ability to invoke the
  Lamda in the same region where Kong is located.
  [#7765](https://github.com/Kong/kong/pull/7765)
- **Datadog**: `host` and `port` config options can be configured from environment variables
  `KONG_DATADOG_AGENT_HOST` and `KONG_DATADOG_AGENT_PORT`. This allows to set different
  destinations on a per Kong node basis, which makes multi-DC setups easier and in Kubernetes allows to
  run the datadog agents as a daemon-set.
  [#7463](https://github.com/Kong/kong/pull/7463)
  Thanks [rallyben](https://github.com/rallyben) for the patch!
- **Prometheus:** A new metric `data_plane_cluster_cert_expiry_timestamp` is added to expose the Data Plane's cluster_cert expiry timestamp for improved monitoring in Hybrid Mode. [#7800](https://github.com/Kong/kong/pull/7800).

> 

Reviewer: ```suggestion
**Request Termination**: New `trigger` config option, which makes the plugin only activate for any requests with a header or query parameter named like the trigger. This can be a great debugging aid, without impacting actual traffic being processed. [#6744](https://github.com/Kong/kong/pull/6744). The `request-echo` config option was added. If set, the plugin responds with a copy of the incoming request. This eases troubleshooting when Kong is behind one or more other proxies or LB's, especially when combined with the new 'trigger' option. [#6744](https://github.com/Kong/kong/pull/6744).

```

Reviewer: would need formatting.. posted on a single line..

**GRPC-Gateway**:

- Fields of type `.google.protobuf.Timestamp` on the gRPC side are now
  transcoded to and from ISO8601 strings in the REST side.
  [#7538](https://github.com/Kong/kong/pull/7538)
- URI arguments like `..?foo.bar=x&foo.baz=y` are interpreted as structured
  fields, equivalent to `{"foo": {"bar": "x", "baz": "y"}}`
  [#7564](https://github.com/Kong/kong/pull/7564)
  Thanks [git-torrent](https://github.com/git-torrent) for the patch!

> **Request Termination**:
> 
> - New `trigger` config option, which makes the plugin only activate for any requests with a header or query parameter
>   named like the trigger. This can be a great debugging aid, without impacting actual traffic being processed.
>   [#6744](https://github.com/Kong/kong/pull/6744).
> - The `request-echo` config option. If set, the plugin responds with a copy of
>   the incoming request. This eases troubleshooting when Kong is behind one or more
>   other proxies or LB's, especially when combined with the new 'trigger' option.
>   [#6744](https://github.com/Kong/kong/pull/6744).
> 

Reviewer: ```suggestion
```
